### ISSUE TITLE
Lint on broken recursion structure

### ISSUE THREAD

**Issue by lf-** [NixOS/nixpkgs#312345]
## Issue description

It's possible to write something like:

```nix
{somePackageName, stdenv}:
stdenv.mkDerivation {
  pname = "some-package-name";
  # ...
  passthru.someOtherThing = pkgs.callPackage ./something.nix { inherit somePackageName; };
}
```

where something in `passthru` doesn't follow the containing derivation's `overrideAttrs` properly due to broken recursion structure. The correct code here would be:

```nix
{somePackageName, stdenv}:
stdenv.mkDerivation (finalAttrs: {
  pname = "some-package-name";
  # ...
  passthru.someOtherThing = pkgs.callPackage ./something.nix { somePackageName = finalAttrs.finalPackage; };
})
```

I would like to have an automated lint that rejects code that exhibits this broken recursion structure: `passthru` referring to the parent derivation but the passthru attrs get the parent derivation from other means than `finalAttrs.finalPackage`.

**Comment by eclairevoyant** [comment]
Why? Also, shouldn't we ensure all the builders support the `finalAttrs` pattern in the first place?

**Comment by lf-** [comment]
> Also, shouldn't we ensure all the builders support the `finalAttrs` pattern in the first place?

yeah, there's a lot of very broken builders that have very broken overrides in general. we should at least fix the standard mkDerivation though.

**Comment by lf-** [comment]
not stale! bad bot

**Comment by philiptaron** [comment]
Now that extendMkDerivation has landed, and the various build helpers are all gaining good overrideAttrs and finalAttrs support, I think nixpkgs-vet is now in a fine place to make this sort of assertion.
